Add ismip6 retreat parameterization

trhille opened the pull request:
This PR adds the ISMIP6 glacier retreat parameterization described by Slater et al. (2019, 2020), which specifies a retreat length ∆L = k * ∆(Q^0.4 * TF), where Q is the mean summer discharge from the subglacial hydrologic system, TF is ocean thermal forcing, and k is a tuning parameter. This is activated using the namelist option config_calving = 'ismip6_retreat'. For now, the tuning parameter k is set using config_ismip6_retreat_k. In the future, this could be replaced by a 2D field in an input stream, as k likely needs to be tuned individually by glacier or ice-sheet sector. The glacier margin can advance if
 ∆(Q^0.4 * TF) is negative due to a colder ocean and/or less subglacial discharge compared with previous year, but the rate of advance can never exceed ice velocity (i.e., there is no freeze-on).

Because this parameterization requires two time-levels from ismip6_2dThermalForcing and ismip6Runoff, this PR includes changes to the MPAS framework that allow easier manipulation of multiple times by adding a mostRecentAccessTime attribute to streams.

I tested this on the Humboldt_1to10km_r04_20210615 mesh using MIROC5 RCP8.5 climate, q = 1/5 basal friction law exponent, and three different values of k corresponding to the 25th, 50th, and 75th percentile values from Slater et al. (2019): -60, -170, and -370. The screenshot below shows ice thickness for k=-370, with ice fronts for k=-60 and k=-170 in white and cyan, respectively. Black contour is initial ice extent:
![image](https://user-images.githubusercontent.com/17446278/148982273-57dc3539-6fe8-480b-93f2-4de2e6248da8.png)

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,230 @@ end subroutine von_Mises_calving
+                  ! We found the stream we are looking for
+                  if (alarm_cursor % isRecurring) then
+                     ! Force a read of this stream, and use the most recent time available in the file.
+                     ! TODO: Determine whether there is a smarter way to do this that doesn't require 
```
I agree we should try to fix having this read every timestep.  In some cases it will also be read by the primary streams read (when it's alarm is ringing), so it will end up getting read twice on those timesteps.

I was imagining we could make use of the new `mostRecentAccessTime` attribute to avoid needing to have your retreat routine do any i/o itself.  I think that could look like this:
```
! t1, t2 are times for the older and newer time levels
! TF1, TF2 are thermal forcing fields from those corresponding time levels, and are separate from the TFocean field read by i/o.  They could either be new model variables in Registry or local arrays.  If local, then they both need the 'save' attribute.  Probably more MPAS-y to make them new variables in that package.
if (t2 /= stream_cursor%mostRecentAccessTime) then
   TF1 = TF2
   t1 = t2
   TF2 = TFocean  ! (RHS being the variable actually in streams)
   t2 = stream_cursor%mostRecentAccessTime
   ! Same thing for Q
endif
```

One issue with this is you will need a separate init routine to handle getting these fields set up properly the first time.  In that one you might need to force one read to get the previous time level (in case one exists).

trhille replied:
Resolved in major rewrite (mostly commit [16b1883](https://github.com/MALI-Dev/E3SM/pull/18/commits/16b18835a5a5f2f76486b8cb43eedf7d3a79799d))

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,230 @@ end subroutine von_Mises_calving
+                                         * TFoceanPrevious(iCell) ) ) / deltatForcing + &
+                                         sqrt(xvelmean(iCell)**2.0_RKIND + yvelmean(iCell)**2.0_RKIND), 0.0_RKIND)
+         enddo
+      endif
```
Maybe add an `else` here with an error in case the unexpected happens and we missed some logic.

trhille replied:
Added in commit b14ed9b90d74105c801f76be29f8ce5d521695c2

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,230 @@ end subroutine von_Mises_calving
+      ! if somehow deltatForcing = 0.0.
+      if (forcingTime .ne. forcingTimeOld) then
+         do iCell=1, nCells
+            ! For now, hard-coding factor of 4.0 to roughly convert mean annual to mean summer runoff.
```
Isn't mean summer runoff what ISMIP6 provides in their forcing datasets?  The two Slater papers both refer to mean summer, so I would have thought that was what was provided (and hence what we are using).

trhille replied:
The forcing datasets are annual runoff, since they are intended to be used for the submarine melt parameterization.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,230 @@ end subroutine von_Mises_calving
+            ! Factor of 1000.0 is freshwater density.
+            calvingVelocity(iCell) = max(-1.0_RKIND * config_ismip6_retreat_k * ( ( (4.0_RKIND * submergedArea(iCell) &
+                                         * ismip6Runoff(iCell) / 1000.0_RKIND)**0.4_RKIND * TFocean(iCell) ) -  &
+                                         ( (4.0_RKIND * submergedArea(iCell) * ismip6RunoffPrevious(iCell) / 1000.0_RKIND)**0.4_RKIND &
```
Is `submergedArea` necessary?  The retreat param uses Q in m^3/s - isn't that the units used by the forcing dataset?

trhille replied:
Unfortunately ISMIP6 provides forcings in [kg m^-2 s^-1], which is what is used for the submarine melt parameterization.

trhille replied:
Added TO-DO in comment to ensure that we don't miss this when using fields passed from the E3SM coupler.

matthewhoffman reviewed (changes requested):
I added a few questions and made a couple implementation suggestions.  Regarding the framework changes that we want to have in a separate PR, I think they are what we want in any case.

trhille commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,230 @@ end subroutine von_Mises_calving
+               cycle
+            endif
+
+            ! Only force a read of streams that are recurring.  To find out if a stream is recurring, we need to
```
Yes, this is a holdover after copy-pasting your code from https://github.com/MALI-Dev/E3SM/blob/ed19e594592d013f3d374a01e6a2927f1ce486de/components/mpas-albany-landice/src/mode_forward/mpas_li_core.F#L174-L194
Testing confirms that this isn't necessary, so I will remove the check for the recurring alarm.

trhille replied:
Removed in commit 8f98bac986efa39a68f95866fc0a5aa740a00e4a

trhille commented:
@matthewhoffman, commit 16b1883 removes unnecessary I/O, as requested. This still uses the saveActualWhen = .true. argument, which we have decided to remove soon. Testing indicates that this functions identically to commit 0d1f477:
 Blue is using commit 16b1883, orange is with commit 0d1f477:
![image](https://user-images.githubusercontent.com/17446278/152069356-6fb3da83-15ec-4f70-bdfd-aaa1eee830cf.png)

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,230 @@ end subroutine von_Mises_calving
+               call mpas_set_time(forcingTime, dateTimeString=forcingTimeStamp, ierr=err_tmp)
+               call mpas_set_time(forcingTimeOld, dateTimeString=forcingTimeOldStamp, ierr=err_tmp)
+               ! reset mostRecentAccessTime to enable elseif block below
+               call mpas_set_time(stream_cursor%mostRecentAccessTime, dateTimeString=forcingTimeStamp, ierr=err_tmp)
```
I don't think we ever want to manually set `stream_cursor%mostRecentAccessTime`.  We should consider it like read-only and use a local (or module) variable with the `save` attribute to do any bookkeeping.

trhille replied:
Okay, I removed this in commit 01db2f9. It does not change functionality.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,227 @@ end subroutine von_Mises_calving
+               ismip6RunoffCurrent = ismip6Runoff
+
+               ! Force a read of this stream, and use the second most recent time in the file
+               call mpas_stream_mgr_read(domain % streamManager, streamID = stream_cursor % name, rightNow = .true., &
```
It dawned on me that this could be dangerous of any other fields are present in that input stream.  (e.g. you could accidentally clobber `sfcMassBal` with a value for an older time).
Unfortunately, I don't think there is an easy way in streams to know what fields got read or are even present in a stream, so we might have to rely on the user ensuring that only those two fields are in this stream.  Maybe we want to make the prescribed name more specific in that case, e.g. "ismip6-gis-retreat"?

Actually, what we should do here is create an immutable stream in Registry that defines the name and contents of this stream.  That would solve almost the entire problem.  From the Users's Guide, Chapter 5:

> Two classes of streams exist in MPAS: immutable streams and mutable streams. Immutable streams are those for which the set of fields that belong to the stream may not be modified at model run-time; however, it is possible to modify the interval at which the stream is read or written, the filename template describing the files containing the stream on disk, and several other parameters of the stream. In contrast, all aspects of mutable streams, including the set of fields that belong to the stream, may be modified at run-time. The motivation for the creation of two stream classes is the idea that an MPAS core may not function correctly if certain fields are not read in upon model start-up or written to restart files, and it is therefore not reasonable for users to modify this set of required fields at run-time. An MPAS core developer may choose to implement such streams as immutable streams. Since fields may not be added to an immutable stream at run-time, new immutable streams may not be defined at run-time, and the only type of new stream that may be defined at run-time is the mutable stream type.

If the details of how to do that are unclear, let me know.  I totally forgot that this sort of situation is in part why immutable streams were created in the first place.  (The restart stream being the only one we've ever actually used thus far.)

trhille replied:
This is a great idea. I updated this in commit 08fd344. Testing indicates identical results to previous setup.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,227 @@ end subroutine von_Mises_calving
+
+               forcingTimeOld = forcingTime
+               forcingTime = stream_cursor % mostRecentAccessTime
+               call mpas_get_time(forcingTime, dateTimeString=forcingTimeStamp, ierr=err_tmp)
```
Add an `err = ior(err, err_tmp)` line after this line too.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,227 @@ end subroutine von_Mises_calving
+         if (trim(stream_cursor % name) == 'ismip-gis') then
+            if (xtime == simulationStartTime) then
+               ! Use forcing fields and time from most recent time in file, which was read on init.
+               call mpas_get_time(stream_cursor%mostRecentAccessTime, dateTimeString=forcingTimeStamp, ierr=err_tmp)
```
add `err = ior(err, err_tmp)` line

trhille replied:
Added in [ac30f9e](https://github.com/MALI-Dev/E3SM/pull/18/commits/ac30f9ebe9e2ef395060ffe1ef3c2ab4694cc13f).

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,227 @@ end subroutine von_Mises_calving
+               ! sum up length of edges adjacent to open ocean or non-dynamic cells
+               if ( (.not. li_mask_is_dynamic_ice(jCell)) &
+                       .and. (bedTopography(jCell) < seaLevel) ) then
+                  submergedArea(iCell) = submergedArea(iCell) + dvEdge(iEdge) * waterDepth
```
These sorts of calculations really make you want a level-set, don't they? :)

trhille replied:
Someday!

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,227 @@ end subroutine von_Mises_calving
+
+      ! Get the time between forcings to calculate retreat rate.
+      call mpas_get_timeInterval(forcingTime - forcingTimeOld, dt=deltatForcing)
+      call mpas_log_write("deltatForcing = $r", realArgs = (/deltatForcing/))
```
```suggestion
      call mpas_log_write("ismip6 retreat deltatForcing = $r", realArgs = (/deltatForcing/))
```

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,227 @@ end subroutine von_Mises_calving
+      character(len=StrKIND) :: forcingTimeStamp, forcingTimeOldStamp
+      type (MPAS_stream_list_type), pointer :: stream_cursor
+
+      applyToGrounded = .true.
```
It would be slightly simpler and easier to read if you eliminate these local variables and just pass, e.g., `applyToGrounded = .true.` below in the call to `li_apply_front_ablation_velocity`.  But this isn't wrong, so your call.

trhille replied:
Added in [ac30f9e](https://github.com/MALI-Dev/E3SM/pull/18/commits/ac30f9ebe9e2ef395060ffe1ef3c2ab4694cc13f).

trhille commented on components/mpas-albany-landice/src/Registry.xml:
```diff
@@ -138,7 +138,7 @@
 		<nml_option name="config_calving" type="character" default_value="none" units="unitless"
 		            description="Selection of the method for calving ice (as defined below)."
-		            possible_values="'none', 'floating', 'topographic_threshold', 'thickness_threshold', 'mask', 'eigencalving', 'specified_calving_velocity', 'von_Mises_stress', 'damagecalving'"
+		            possible_values="'none', 'floating', 'topographic_threshold', 'thickness_threshold', 'mask', 'eigencalving', 'specified_calving_velocity', 'von_Mises_stress', 'damagecalving', 'ismip6_retreat'"
```
Added in [ac30f9e](https://github.com/MALI-Dev/E3SM/pull/18/commits/ac30f9ebe9e2ef395060ffe1ef3c2ab4694cc13f).
https://github.com/trhille/E3SM/blob/add_ismip6_retreat_parameterization/components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F#L259-L270

trhille commented on components/mpas-albany-landice/src/Registry.xml:
```diff
@@ -240,6 +240,10 @@
 		/>
+                <nml_option name="config_ismip6_retreat_k" type="real" default_value="-170.0" units="m (m^{3} s^{-1})^{-0.4} deg. C^{-1}"
+                            description="Coefficient for ISMIP6 retreat parameterization from Slater et al. (2019)"
+                            possible_values="any negative real value"
```
Added in [ac30f9e](https://github.com/MALI-Dev/E3SM/pull/18/commits/ac30f9ebe9e2ef395060ffe1ef3c2ab4694cc13f).
https://github.com/trhille/E3SM/blob/add_ismip6_retreat_parameterization/components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F#L1771-L1774

trhille commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,227 @@ end subroutine von_Mises_calving
+
+               ! Force a read of this stream, and use the second most recent time in the file
+               call mpas_stream_mgr_read(domain % streamManager, streamID = stream_cursor % name, rightNow = .true., &
+                                when = forcingTimeStamp, whence = MPAS_STREAM_LATEST_STRICTLY_BEFORE, saveActualWhen= .true., ierr=err_tmp)
```
Added in [ac30f9e](https://github.com/MALI-Dev/E3SM/pull/18/commits/ac30f9ebe9e2ef395060ffe1ef3c2ab4694cc13f).

trhille commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,227 @@ end subroutine von_Mises_calving
+               call mpas_stream_mgr_read(domain % streamManager, streamID = stream_cursor % name, rightNow = .true., &
+                                when = forcingTimeStamp, whence = MPAS_STREAM_LATEST_STRICTLY_BEFORE, saveActualWhen= .true., ierr=err_tmp)
+
+               call mpas_get_time(stream_cursor%mostRecentAccessTime, dateTimeString=forcingTimeOldStamp, ierr=err_tmp)
```
Added in [ac30f9e](https://github.com/MALI-Dev/E3SM/pull/18/commits/ac30f9ebe9e2ef395060ffe1ef3c2ab4694cc13f).

trhille commented on components/mpas-albany-landice/src/Registry.xml:
```diff
@@ -701,6 +705,10 @@
 			<var name="ismip6Runoff" packages="ismip6GroundedFaceMelt" />
 			<var name="ismip6_2dThermalForcing" packages="ismip6GroundedFaceMelt" />
+                        <var name="ismip6RunoffPrevious" packages="ismip6GroundedFaceMelt" />
+			<var name="ismip6_2dThermalForcingPrevious" packages="ismip6GroundedFaceMelt" />
```
Testing indicates that they are not. Removed in commit [08fd344](https://github.com/trhille/E3SM/commit/08fd34409b2ac17b4539e02f1a67b50473cf270b).

trhille commented on components/mpas-albany-landice/src/Registry.xml:
```diff
@@ -787,7 +795,11 @@
                         <!-- these variables just for ISMIP6 grounded glacier forcing -->
                         <var name="ismip6Runoff" packages="ismip6GroundedFaceMelt" />
-                        <var name="ismip6_2dThermalForcing" packages="ismip6GroundedFaceMelt" />
+			<var name="ismip6_2dThermalForcing" packages="ismip6GroundedFaceMelt" />
```
Okay, commit 0e38b727d882284b7d0d71df7f3c518b93d717bd fixes the issue on restarts. I did a manual restart test with two runs: one without a restart (blue) and one with a restart after year 5 (orange):
![image](https://user-images.githubusercontent.com/17446278/153520371-0a972186-4c5e-40c3-b21b-e7d4442bd3ef.png)

trhille commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1684,227 @@ end subroutine von_Mises_calving
+
+      ! Update halos on calvingThickness before applying it.
+      ! NOTE: THIS WILL NOT WORK ON MULTIPLE BLOCKS PER PROCESSOR
+      call mpas_timer_start("halo updates")
```
I just did a manual decomposition test with 180 cores (blue) and 96 cores (orange):
![image](https://user-images.githubusercontent.com/17446278/153522114-ffba44ac-e1e5-4be0-b65a-f8cf49e69fe6.png)

These differences seem pretty large, which is worrying.

matthewhoffman replied:
Note for posterity: we determined the decomp BFB error is actually in `li_apply_front_ablation_velocity`.  Fixing that will be saved for a separate PR.

trhille commented:
@matthewhoffman I've addressed all your comments. Let I'm not sure what to think about the results of my decomposition test. That might require some further work.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1836,6 +1861,12 @@ subroutine ismip6_retreat(domain, err)
 
+      ! check that the ismip6-gis stream was found
+      if (.not. streamFound) then
+         call mpas_log_write('Error: Input stream ismip6-gis is required for config_calving = ismip6_retreat, ' // &
+                             'but was not found.')
```
Remove "Error:", and add the `MPAS_LOG_ERR` argument after the string.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -1679,6 +1692,265 @@ end subroutine von_Mises_calving
+      stream_cursor => domain % streamManager % streams % head
+      do while (associated(stream_cursor))
+         ! Input stream with forcings must be called 'ismip6-gis', or this will throw an error.
+         if (trim(stream_cursor % name) == 'ismip6-gis') then
```
```suggestion
         if ((trim(stream_cursor % name) == 'ismip6-gis') .and. &
              (stream_cursor % active_stream)) ) then
```

I think we need this change to prevent the code from detecting the immutable stream definition if that stream itself is not included in `streams.landice`.

matthewhoffman reviewed (approved):
@trhille , thanks for resolving all the remaining issues.  Everything looks good now.  I've tested with the new COMPASS calving tests for the Humboldt mesh I have been working on.  Specifically, the version tested was:
https://github.com/MPAS-Dev/compass/pull/318
at commit https://github.com/MPAS-Dev/compass/pull/318/commits/015cefcecefff07471dcf16a97e6f70e1242154a .

This branch fails the decomposition test `landice/humboldt/mesh-default_decomposition_test/velo-none_calving-ismip6_retreat`, as expected due to the known bug in `li_apply_front_ablation_velocity`, with the following reasonable values:
```
Running steps: 1proc_run, 32proc_run
  * step: 1proc_run
  * step: 32proc_run
thickness            Time index: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16
1:  l1: 4.77117463077826e+01  l2: 7.01525349523819e+00  linf: 2.84146164592161e+00
2:  l1: 1.04110458582546e+02  l2: 1.87809816517174e+01  linf: 1.09631805085328e+01
3:  l1: 2.26255993273168e+02  l2: 5.37776670577194e+01  linf: 4.04141293730310e+01
4:  l1: 3.25257667936368e+02  l2: 8.46968455544373e+01  linf: 5.68885920198856e+01
5:  l1: 3.71501737217254e+02  l2: 6.58581107960222e+01  linf: 3.35737945050176e+01
6:  l1: 4.77961032770880e+02  l2: 7.52411221845642e+01  linf: 3.55750552527465e+01
7:  l1: 5.86111399818751e+02  l2: 1.03063155913136e+02  linf: 4.61264205227019e+01
8:  l1: 4.46555633917253e+02  l2: 6.88606163937160e+01  linf: 3.28648004786271e+01
9:  l1: 4.48735820084300e+02  l2: 7.14231629713249e+01  linf: 3.54428946950350e+01
10:  l1: 4.25741344827961e+02  l2: 6.29966384583536e+01  linf: 3.09523248890635e+01
11:  l1: 4.14149644691802e+02  l2: 6.14411868614030e+01  linf: 3.04149536266058e+01
12:  l1: 5.76201937260433e+02  l2: 9.62223314192466e+01  linf: 5.79610555617643e+01
13:  l1: 7.67888656741156e+02  l2: 1.26511583263426e+02  linf: 5.22280556810034e+01
14:  l1: 5.16813430485597e+02  l2: 8.93237173153720e+01  linf: 5.92944551642646e+01
15:  l1: 6.11877268577072e+02  l2: 1.02648842098631e+02  linf: 4.62382486726791e+01
16:  l1: 6.67152888605527e+02  l2: 1.04084789967847e+02  linf: 5.50468617233342e+01
  FAIL /lustre/scratch4/turquoise/mhoffman/junk/landice/humboldt/mesh-default_decomposition_test/velo-none_calving-ismip6_retreat/1proc_run/output.nc

       /lustre/scratch4/turquoise/mhoffman/junk/landice/humboldt/mesh-default_decomposition_test/velo-none_calving-ismip6_retreat/32proc_run/output.nc

calvingThickness     Time index: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16
1:  l1: 4.77117463077828e+01  l2: 7.01525349523822e+00  linf: 2.84146164592163e+00
2:  l1: 5.81128792018471e+01  l2: 1.05766404085892e+01  linf: 6.87299755880575e+00
3:  l1: 1.76411956635276e+02  l2: 4.68645314890459e+01  linf: 3.27263724944545e+01
4:  l1: 3.43961624046429e+02  l2: 6.57180188991484e+01  linf: 4.43369179601394e+01
5:  l1: 4.24549190495720e+02  l2: 7.65385776846250e+01  linf: 3.22007478376397e+01
6:  l1: 2.92204819505360e+02  l2: 4.71358697772873e+01  linf: 2.41462225895031e+01
7:  l1: 3.33922262172565e+02  l2: 7.35386516397582e+01  linf: 4.08544435549891e+01
8:  l1: 2.98697832998599e+02  l2: 7.43244723640048e+01  linf: 4.36215681873039e+01
9:  l1: 1.86015927785989e+02  l2: 3.82241806207109e+01  linf: 2.65646317574508e+01
10:  l1: 1.79285479739526e+02  l2: 4.76760270702290e+01  linf: 2.92433408335022e+01
11:  l1: 1.23013234144608e+02  l2: 2.41416848689712e+01  linf: 1.31873532537006e+01
12:  l1: 2.24630271687293e+02  l2: 5.04964778760934e+01  linf: 3.22400245021401e+01
13:  l1: 3.98650450903705e+02  l2: 7.38650945175085e+01  linf: 3.38733450832966e+01
14:  l1: 4.81838558145487e+02  l2: 8.65689034529773e+01  linf: 3.61682053305072e+01
15:  l1: 2.93214111079274e+02  l2: 6.38441638946028e+01  linf: 3.75448960724951e+01
16:  l1: 5.17726816709688e+02  l2: 8.47132894365908e+01  linf: 3.77746890136409e+01
  FAIL /lustre/scratch4/turquoise/mhoffman/junk/landice/humboldt/mesh-default_decomposition_test/velo-none_calving-ismip6_retreat/1proc_run/output.nc

       /lustre/scratch4/turquoise/mhoffman/junk/landice/humboldt/mesh-default_decomposition_test/velo-none_calving-ismip6_retreat/32proc_run/output.nc

```

It passes the restart test `landice/humboldt/mesh-default_restart_test/velo-none_calving-ismip6_retreat`:
```
Running steps: full_run, restart_run
  * step: full_run
  * step: restart_run
thickness            Time index: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16
  PASS /lustre/scratch4/turquoise/mhoffman/junk/landice/humboldt/mesh-default_restart_test/velo-none_calving-ismip6_retreat/full_run/output.nc

       /lustre/scratch4/turquoise/mhoffman/junk/landice/humboldt/mesh-default_restart_test/velo-none_calving-ismip6_retreat/restart_run/output.nc

normalVelocity       Time index: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16
  PASS /lustre/scratch4/turquoise/mhoffman/junk/landice/humboldt/mesh-default_restart_test/velo-none_calving-ismip6_retreat/full_run/output.nc

       /lustre/scratch4/turquoise/mhoffman/junk/landice/humboldt/mesh-default_restart_test/velo-none_calving-ismip6_retreat/restart_run/output.nc

calvingThickness     Time index: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16
  PASS /lustre/scratch4/turquoise/mhoffman/junk/landice/humboldt/mesh-default_restart_test/velo-none_calving-ismip6_retreat/full_run/output.nc

       /lustre/scratch4/turquoise/mhoffman/junk/landice/humboldt/mesh-default_restart_test/velo-none_calving-ismip6_retreat/restart_run/output.nc
```

trhille commented:
@matthewhoffman, Thanks for such careful reviews and testing! This passed all baseline comparisons in COMPASS testing.
